lora: Use generic node name "lora" for all lora nodes

Mani-Sadhasivam opened the pull request:
The current devicetree specification (master branch [1]) lists "lora" as a
generic node name. Hence, all devicetree files in Zephyr should use the same.
    
~~As a result of this change, we no longer need an alias for LoRa and the
applications can now use DT_NODELABEL for getting the node identifiers instead
of DT_ALIAS.~~
    
[1] https://github.com/devicetree-org/devicetree-specification/blob/main/source/chapter2-devicetree-basics.rst
    
Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

mniestroj commented on boards/arm/96b_wistrio/96b_wistrio.dts:
```diff
@@ -86,7 +85,7 @@
 	cs-gpios = <&gpiob 0 GPIO_ACTIVE_LOW>;
 
-	lora: sx1276@0 {
+	lora0: lora@0 {
```
This commit does two things:
- change node name from something like `sx1276` / `sx1262` / `radio` to `lora`,
- drop `/aliases/lora0` in favour of `lora0` nodelabel.

The first change in justified in commit message. However the second change does't seem to be related to the first and I also do not see the reason why we should use nodelabel instead of an alias. I'd like second commit message for the second change, with a good reason in commit body why we want that.

Mani-Sadhasivam replied:
Hmm, since those two changes were interdependent I thought having them together will help in bisecting. But I don't have a strong opinion either, so will split it up.

mniestroj replied:
It should be:
```
lora: lora@0 {
```
in this commit, so that `/aliases/lora0` is still valid.

Mani-Sadhasivam committed:
lora: Use generic node name "lora" for all lora nodes

The current devicetree specification (master branch [1]) lists "lora" as
a generic node name. Hence, all devicetree files in Zephyr should use
the same.

[1] https://github.com/devicetree-org/devicetree-specification/blob/main/source/chapter2-devicetree-basics.rst

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

henrikbrixandersen reviewed (commented):
I suggest going with a `zephyr,lora` `chosen` node instead, similar to how other subsystems handle default devices. See https://docs.zephyrproject.org/latest/reference/devicetree/api.html#devicetree-chosen-nodes for more information and examples.

thedjnK commented:
If this method changes, please ensure the arm/rm1xx_dvk board is updated also

mbolivar-nordic reviewed (changes requested):
The first patch in the current revision is fine. That's just making the devicetrees follow along with the current specification by renaming some nodes -- no issues there.

I've got some issues with the second patch, one simple and one not so simple.

1. Simple: Please don't change DEVICE_DT_GET() to device_get_binding(). That's adding unnecessary overhead. I'm not asking you to change device_get_binding() to DEVICE_DT_GET(), although it would improve the code if you did. I am asking you not to change DEVICE_DT_GET() to device_get_binding(), because it makes the code worse.

2. Not-so-simple: it's not clear to me why the alias is bad and why you want to get rid of it, but I agree that current node label is a bit weird.

Zephyr's use of aliases and chosen nodes is documented here: https://docs.zephyrproject.org/latest/guides/dts/intro.html#aliases-and-chosen-nodes.

The second patch's C changes are almost entirely in the sample code, which would point towards just keeping this an alias. However, it also touches the lora shell, which as @henrikbrixandersen says is more along the lines of what we typically use chosen nodes for, since it's part of generic driver code.

However, we honestly don't have a good answer for how shell code can find devices of a particular "type". For example, drivers/adc/adc_shell.c has a big hacky pile of ifdefs depending on the SoC, because there's no way to iterate over existing devices and asking what type of API they implement.

And I think a chosen node is probably a bad idea if you are only going to use this in samples and the shell. The shell is mostly just for debugging and testing; a `zephyr,foo` chosen node is typically a way to point to a device that some common subsystem code needs to use in a "real application".

We had some discussion with @galak a while ago about adding type-specific device constructors that would say "this is an ADC device, not any other kind of device". We still don't have that for ADC devices, but we do now have `I2C_DEVICE_DT_INST_DEFINE` etc. for I2C devices.

If you are looking for a generic solution for the lora shell to find lora nodes that isn't an alias, then I would suggest adding something similar for lora, `LORA_DEVICE_DT_INST_DEFINE` or so. Then we could have CONFIG_LORA_SHELL `select` another Kconfig option that forces `LORA_DEVICE_DT_INST_DEFINE` to also add an entry to a little array of LORA device handles using linker magic. Then the shell could use that array to find lora devices.

If this is just for the shell and the current hack with an alias is not good enough for some good reason, I think that's probably the way to go. I think moving to a node label would just be trading one hack for another, and I'm not sure why that would be any better -- can you clarify?

On the other hand, if you are going to follow up this pull request with other patches that need to find "the system lora device" in generic subsystem code besides the shell, then a chosen node is definitely the way to go.

Mani-Sadhasivam commented:
> If this is just for the shell and the current hack with an alias is not good enough for some good reason, I think that's probably the way to go. I think moving to a node label would just be trading one hack for another, and I'm not sure why that would be any better -- can you clarify?

The way I envision `alias` was providing a common name for nodes that have different vendor/model specific node labels. For instance, `zephyr,console` alias can be defined for SERIAL device nodes such as `uart`, `lpuart`, etc... In this case, `alias` makes perfect sense as we are declaring what SERIAL device should be used as a Zephyr console since we have different kinds of them in a system.

But for the case of lora, we are going to use `lora` as the node label for all LoRa based nodes due to the 1st commit in this PR. So here, defining the alias for the already generic node label seemed overkill to me. So that's why I thought we can directly use the node label to get the device in both sample and shell.

But I should admit that I missed reading Zephyr's documentation about alias and chosen nodes earlier.

mbolivar-nordic commented:
Happy new year! @Mani-Sadhasivam I'm not sure how to understand your reply. I think you may have been mixing up 'alias' and 'chosen node' in some places. But it sounds to me like you're going to be using a system-wide lora device at some point in a bigger way in later patches, right? If that's the case, I think a chosen node is the usual zephyr way of doing things.

Mani-Sadhasivam commented:
@mbolivar-nordic I've removed the commit that got rid of the alias. Regarding using chosen nodes, I think we do not need it _yet_. Please review!

erwango commented on boards/arm/96b_wistrio/96b_wistrio.dts:
```diff
@@ -90,7 +90,7 @@
 	cs-gpios = <&gpiob 0 GPIO_ACTIVE_LOW>;
 
-	lora: sx1276@0 {
+	lora: lora@0 {
```
Is there a need to keep the first lora occurence now ?

Mani-Sadhasivam replied:
Sorry, I don't get what you are saying.

henrikbrixandersen replied:
The first is the nodelabel, the second is the node name. They are both needed.

erwango commented on dts/arm/st/wl/stm32wl.dtsi:
```diff
@@ -292,7 +292,7 @@
 			use-subghzspi-nss;
 
-			radio@0 {
+			lora@0 {
```
This doesn't feel right, this radio can be used for other purposes than lora ((G)FSK, (G)MSK, and BPSK.)

Mani-Sadhasivam replied:
This is a Lora device that can work in multiple modulation modes you mentioned. In all the cases, it is sold as a "LoRa module" only, so the node name matches.

erwango replied:
I disagree, this node is common to the whole stm32wl seriers.
Looking for instance to the STM32WLE ref manual:
Headline: 
```
STM32WLEx advanced Arm ® -based 32-bit MCUs with sub-GHz radio solution
```
Description: 
```
STM32WLEx microcontrollers with integrated sub-GHZ radio operating in the
150 - 960 MHz ISM band, belong to a family of microcontrollers with different memory sizes,
packages and peripherals.
```
Radio subystem description:
```
4.5 Sub-GHz radio modems
The following modems are provided enabling the respective modulation and associated
framing.
• LoRa with the LoRa framing
• FSK and MSK with a generic framing
• BPSK with a BPSK framing
```

Lora is clearly a one of multiple modems, not the only one

Mani-Sadhasivam replied:
Ah, sorry. I overlooked the STM32WL part. This is clearly different from the LoRa modules available in the market as it supports multiple modulations in a single chip. But shouldn't this be documented in devicetree spec?

Mani-Sadhasivam commented:
@erwango @mbolivar-nordic Please review!

Mani-Sadhasivam commented:
@carlescufi Please merge!
